### PR TITLE
`register_artifact ` function

### PR DESCRIPTION
**Pull request by avishniakov**
## Describe changes
I implemented `register_artifact ` function to support use-cases when users want to just link already existing data inside the artifact store scope manually saved by some other tools.

Docs section has quite some examples of the usage in various scenarios.

Example use-case: Pytorch-Lightning has the ability to store checkpoints on a given remote storage, so there is no need to do `save_artifact` in the end epoch callback or similar, but rather just link existing checkpoint to be available as a ZenML artifact for the future use.

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).
- [x] If my changes require changes to the dashboard, these changes are communicated/requested.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: docs/book/user-guide/starter-guide/manage-artifacts.md
Comment (htahir1) [inline docs/book/user-guide/starter-guide/manage-artifacts.md:323]: might make sense to somehow use some internal methods to get the bucket from the artifact store, like active_stack.artifact_store.uri because then it can work in the local case as well without hardcoding?

File: docs/book/user-guide/starter-guide/manage-artifacts.md
Comment (htahir1) [inline docs/book/user-guide/starter-guide/manage-artifacts.md:329]: cant we also have callbacks to do this one by one in case training gets interrupted? might make sense to have `link_file_as_artifact` for that

Comment (htahir1) [reply docs/book/user-guide/starter-guide/manage-artifacts.md:329]: How will versioning work in this case?

**Review by htahir1 — CHANGES_REQUESTED** [review]
Left some comments as most people are out today :-)

File: docs/book/user-guide/starter-guide/manage-artifacts.md
Comment (avishniakov) [reply docs/book/user-guide/starter-guide/manage-artifacts.md]: Reworked

File: docs/book/user-guide/starter-guide/manage-artifacts.md
Comment (avishniakov) [reply docs/book/user-guide/starter-guide/manage-artifacts.md]: In this case, the versioning would not work too much, but I added an extra example to this section.
In fact, `link_file_as_artifact` would mean the need to move data around, since our materialized relies on folders, AFAIK. I would wait for @schustmi to confirm on Monday, but even with the current approach on folders, I'm rather satisfied with how this experience could look like under a Callback of Lightning directly.

**Comment by avishniakov** [comment]
@wjayesh since we had a word about this on Discord, you might want to look into this before we merge as well.

File: src/zenml/artifacts/load_directory_materializer.py
Comment (schustmi) [inline src/zenml/artifacts/load_directory_materializer.py:26]: If I remember correctly, I implemented a `DirectoryMaterializer` which works to both save and load directories when an artifact was of type `Path`. Any specific reason we do not implement the `save(...)` method for this materializer as well?

I might be wrong here, but because this materializer has the `pathlib.Path` as it's associated type, it will even be registered for this type in the materializer registry in case it gets imported somehow? Which would then break all returns of type `pathlib.Path`.

File: src/zenml/artifacts/load_directory_materializer.py
Comment (schustmi) [inline src/zenml/artifacts/load_directory_materializer.py:48]: I'm guessing we could add a check that this is indeed a directory, not a file? Otherwise the copying will fail/have unexpected behaviour.

**Review by schustmi — CHANGES_REQUESTED** [review]

File: src/zenml/artifacts/load_directory_materializer.py
Comment (avishniakov) [reply src/zenml/artifacts/load_directory_materializer.py:26]: The `save` was not implemented to protect from misuse of this specific materializer by intent. And this implementation echos the `DirectoryMaterializer` you created in projects.

Yes, it works with `pathlib.Path`, shall I create a child class from `pathlib.Path` to protect this case?

File: docs/book/user-guide/starter-guide/manage-artifacts.md
Comment (schustmi) [reply docs/book/user-guide/starter-guide/manage-artifacts.md]: I'm guessing the callback will create files with the epoch name in the checkpoints directory? In that case, the repeated calls to this method would create artifact versions, but they would all point to the same directory which will have all the checkpoint files inside?

Currently when ZenML creates an artifact, we usually point the path to a directory. However we might be able to workaround that by simply creating an artifact which points to a specific file, as long as the materializer handles it correctly? Not sure if there are additional checks and whether we get the exact checkpoint file URL inside the handler, but worth a try I guess.

If that's not possible, it's somehow not that useful that actually call that at the end of every epoch, right? Just doing it once before the training call would actually be good enough I assume?

Comment (avishniakov) [reply docs/book/user-guide/starter-guide/manage-artifacts.md]: I depicted how this shall be used in the example in docs, but you are all right - pointing to a file is more intuitive for the checkpointing case (and for others maybe). I will try to switch gears here and point to the file instead.

File: src/zenml/artifacts/load_directory_materializer.py
Comment (schustmi) [reply src/zenml/artifacts/load_directory_materializer.py:26]: I see, I think my preferred option would be somehow allowing special materializers without associated types that don't get picked up as default for anything in the registry. I'm not sure however how much effort this is.

Something I think we should do in any case, even if we keep it as-is: The `NotImplementedError` that gets raised in the `save(...)` method should explain that this is a special materializer and should not be used for regular artifacts.

File: docs/book/user-guide/starter-guide/manage-artifacts.md
Comment (avishniakov) [reply docs/book/user-guide/starter-guide/manage-artifacts.md]: I made it available for single files as well

File: src/zenml/artifacts/load_directory_materializer.py
Comment (avishniakov) [reply src/zenml/artifacts/load_directory_materializer.py]: I made it through the `PreexistingArtifactPath` derived from the `Path`, otherwise logic of input artifacts blows up. Please check the recent changes again 🙂

File: src/zenml/artifacts/utils.py
Comment (schustmi) [inline src/zenml/artifacts/utils.py:323]: This should be `PreexistingArtifactPath` I think?

File: src/zenml/artifacts/load_directory_materializer.py
Comment (schustmi) [inline src/zenml/artifacts/load_directory_materializer.py:26]: Does `class PreexistingArtifactPath(Path):` not work for some reason?

File: src/zenml/artifacts/load_directory_materializer.py
Comment (schustmi) [inline src/zenml/artifacts/load_directory_materializer.py:74]: I still think this should have at least a short error message explaining to not use this materializer to save things, just in case

File: src/zenml/artifacts/utils.py
Comment (schustmi) [inline src/zenml/artifacts/utils.py:258]: Maybe we could shorten this to something like `register_artifact(...)`? Or do you think that would not be clear enough in what it does?

**Review by schustmi — CHANGES_REQUESTED** [review]

File: src/zenml/artifacts/load_directory_materializer.py
Comment (avishniakov) [reply src/zenml/artifacts/load_directory_materializer.py:26]: Yeah, it fails to init some internal variables, like `_flavor`. I tried using `super().__init__`, but it didn't help.
Some random guys from the Internet suggest to subclass like this for `Path`.

File: src/zenml/artifacts/utils.py
Comment (avishniakov) [reply src/zenml/artifacts/utils.py]: Yeah, thanks for the catch

File: src/zenml/artifacts/utils.py
Comment (avishniakov) [reply src/zenml/artifacts/utils.py:259]: Doch, `register_artifact(...)` reflects it well, I will switch

File: src/zenml/artifacts/load_directory_materializer.py
Comment (avishniakov) [reply src/zenml/artifacts/load_directory_materializer.py]: Added

File: src/zenml/artifacts/utils.py
Comment (avishniakov) [reply src/zenml/artifacts/utils.py]: Renamed

**Review by schustmi — APPROVED** [review]

**Comment by avishniakov** [comment]
@htahir1 would need you 🦭 to merge this sometime

**Review by htahir1 — CHANGES_REQUESTED** [review]
Some minor do overs but almost there

**Review by bcdurak — CHANGES_REQUESTED** [review]
Left some comments as well.

File: src/zenml/artifacts/load_directory_materializer.py
Comment (schustmi) [inline src/zenml/artifacts/load_directory_materializer.py:26]: @bcdurak and me just looked at this again and thought that having an extra type just for this was still a little weird. What do you think of using `Path` as the `ASSOCIATED_TYPE`, and setting `SKIP_REGISTRATION=True` for this materializer?

This way, it would not be used when users return an artifact of type `Path`, but it will be used to load artifacts that were created by the `register_artifact` function because the materializer is specified explicitly.

Comment (avishniakov) [reply src/zenml/artifacts/load_directory_materializer.py:26]: Yeah, this is great idea

File: src/zenml/artifacts/utils.py
Comment (bcdurak) [inline src/zenml/artifacts/utils.py:661]: I forgot to mention this here. Since you moved the function out, we might want to rephrase the error message here. Previously this was only used in a context where we were trying to create an artifact, now it is a check that can be used in different settings as well.

Comment (avishniakov) [reply src/zenml/artifacts/utils.py:661]: I would rather disagree here: the context is still the same - we try to create AV in ZenML on a given URI and it fails to do so since URI is already taken. I'd prefer to keep this one as is.

File: docs/book/user-guide/starter-guide/manage-artifacts.md
Comment (avishniakov) [reply docs/book/user-guide/starter-guide/manage-artifacts.md:1]: Done so

File: docs/book/user-guide/starter-guide/manage-artifacts.md
Comment (htahir1) [inline docs/book/user-guide/starter-guide/manage-artifacts.md:304]:
```suggestion
### Linking existing data as a ZenML artifact
```

File: src/zenml/__init__.py
Comment (htahir1) [inline src/zenml/__init__.py:40]: extra line?

**Review by htahir1 — APPROVED** [review]
Almost there! but consider it approved from my side once you apply final changes :-) Great job!

File: src/zenml/__init__.py
Comment (avishniakov) [inline src/zenml/__init__.py:40]:
```suggestion
```

**Review by bcdurak — APPROVED** [review]

**Pull request merged**
